aten/src/ATen/native/Gesv.cpp

```diff
@@ -72,6 +78,107 @@ static void applyGesv(Tensor& b, Tensor& A, std::vector<int64_t> infos) {
   }
 }
 
+std::tuple<Tensor&,Tensor&> _gesv_single_out_cpu(
+    Tensor& sol, Tensor& lu,
+    const Tensor& self, const Tensor& A) {
+#ifndef USE_LAPACK
+  AT_ERROR("gesv: LAPACK library not found in compilation");
+#endif
+  int64_t bx = self.size(0);
+  int64_t by = (self.dim() == 1) ? 1 : self.size(1);
+  int64_t ax = A.size(0);
+  int64_t ay = A.size(1);
+  int info = 0;
+  Tensor temp_sol;
+  Tensor temp_lu;
+
+  /* Init to column major format. `sol` and `lu` need to be contiguous
+   * since we pass sol.data() and lu.data() to Lapack */
+  bool tc_sol = isTransposeContiguous(sol);
+  auto self_t = self.view({bx, by}).t();
+  if (!tc_sol &&
+      !sol.is_contiguous() &&
+      sol.dim() == 2 &&
+      sol.size(0) == bx &&
+      sol.size(1) == by) {
+    // eg. torch.gesv(.., out=(n[::2], ...))
+    temp_sol = self_t.clone();
+  }
+
+  if (!temp_sol.defined()) {
+    if (tc_sol) {
+      sol.t().resize_({by, bx});
+    } else {
+      sol.resize_({by, bx});
+    }
+    if (&self == &sol) {
+      // note: not a comprehensive check. Will fail for views
+      if (!tc_sol) {
+        sol.copy_(self_t.clone());
+      }
+    } else {
+      if (tc_sol) {
+        sol.t().copy_(self_t);
+      } else {
+        sol.copy_(self_t);
+      }
+    }
+  }
+
+  bool tc_lu = isTransposeContiguous(lu);
+  if (!tc_lu &&
+      !lu.is_contiguous() &&
+      lu.dim() == 2 &&
+      lu.size(0) == ax &&
+      lu.size(1) == ay) {
+    temp_lu = A.t().clone();
+  }
+
+  if (!temp_lu.defined()) {
+    if (tc_lu) {
+      lu.t().resize_({ay, ax});
+    } else {
+      lu.resize_({ay, ax});
+    }
+    if (&A == &lu) {
+      if (!tc_lu) {
+        lu.copy_(A.t().clone());
+      }
+    } else {
+      if (tc_lu) {
+        lu.t().copy_(A.t());
+      } else {
+        lu.copy_(A.t());
+      }
+    }
+  }
+
+  AT_DISPATCH_FLOATING_TYPES(self.type(), "gesv", [&]{
+    auto A_ptr = temp_lu.defined() ? temp_lu.data<scalar_t>()
+                                   : lu.data<scalar_t>();
+    auto b_ptr = temp_sol.defined() ? temp_sol.data<scalar_t>()
+                                    : sol.data<scalar_t>();
+    auto ipiv = at::empty({bx}, sol.options().dtype(kInt));
+    lapackGesv<scalar_t>(bx, by, A_ptr, bx, ipiv.data<int>(), b_ptr, bx, &info);
+  });
+
+  checkErrors({info});
+
+  if (temp_sol.defined()) {
+    sol.copy_(temp_sol.t_());
+  } else if (!tc_sol) {
+    sol.t_();
```

Comment:
Why do you need this inplace transform here? I think you should just copy into `sol` in a way so that it is column major contiguous already, right?

Reply:
If `tc_sol` or `tc_lu` is true, we're implicitly using their transposed version all along since it is contiguous and we need to give Lapack a pointer to contiguous data. After Lapack has been called, since the original sol/lu already had a transposing view on it, we can just return it. However if `tc_sol`/`tc_lu` is false, then we need to transpose the result from Lapack before returning it.

Reply:
Yes, I agree with you that `t_` is needed with the current code. I was thinking if in the 3rd case you did `sol.resize_({by, bx}).copy_(self.t()).t_();`, the `t_` here isn't needed, and it makes the code easier to read since you have the invariant that no matter which path you take in the big `if` above, `sol` has correct shape. What do you think?

Reply:
Yeah I put it here since we wanted to avoid using t_() before all error checks were done, but I can move it to case 3